libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,759 @@
+                    _max_id = i;
+                }
+            }
+        }
+
+        if (_id_count == 0
+                || _max_id - _min_id > _id_count - 1) { // loop forever
+            _is.wake_from_bl = 1;
+            return _is.active_id;
+        }
+        _fast_throttle_byte_count = 1;
+        int8_t bitCount = 12 + (_id_count * 11);
+        while (bitCount > 0) {
+            _fast_throttle_byte_count++;
+            bitCount -= 8;
+        }
+        _is.set_fast_command[1] = _fast_throttle_byte_count; // just for older ESC FW versions since 1.0 001 this byte is ignored as the ESC calculates it itself
+        _is.set_fast_command[2] = _min_id;                 // min ESC id
+        _is.set_fast_command[3] = _id_count;               // count of ESCs that will get signals
+    }
+
+    if (_is.delay_loops > 0) {
+        _is.delay_loops--;
+        return _setup_active;
+    }
+
+    if (_is.active_id < _setup_active) {
+        _is.active_id = _setup_active;
+        _is.state = 0;
+        _is.timeout = 0;
+    }
+
+    if (_is.timeout == 3 || _is.timeout == 6 || _is.timeout == 9 || _is.timeout == 12) {
+        pull_reset();
+    }
+
+    if (_is.timeout < 15) {
+        if (_is.wake_from_bl) {
+            switch (_is.state) {
+            case 0:request[0] = OW_BL_START_FW;
+                if (_found_escs[_is.active_id].in_boot_loader == 1) {
+                    transmit(_is.active_id, request, _request_length[request[0]]);
+                    _is.delay_loops = 5;
+                } else {
+                    return _is.active_id + 1;
+                }
+                _is.state = 1;
+                break;
+            case 1:request[0] = OW_OK;
+                if (pull_command(_is.active_id, request, response, return_type::FULL_FRAME)) {
+                    _is.timeout = 0;
+                    if (response[0] == 0x02) {
+                        _found_escs[_is.active_id].in_boot_loader = 1;
+                        _is.state = 0;
+                    } else {
+                        _found_escs[_is.active_id].in_boot_loader = 0;
+                        _is.delay_loops = 1;
+                        return _is.active_id + 1;
+                    }
+                } else {
+                    _is.timeout++;
+                }
+                break;
+            }
+        } else {
+            if (pull_command(_is.active_id, _is.set_fast_command, response, return_type::RESPONSE)) {
+                _is.timeout = 0;
+                _is.delay_loops = 1;
+                return _is.active_id + 1;
+            } else {
+                _is.timeout++;
+            }
+        }
+    } else {
+        pull_reset();
+        return _is.active_id + 1;
+    }
+    return _is.active_id;
+}
+
+
+/**
+    increment package count for every ESC
+*/
+void AP_FETtecOneWire::inc_send_msg_count()
+{
+    _send_msg_count++;
+    if (_send_msg_count > 4 * _update_rate_hz) { // resets every four seconds

```

Comment:
timestamp based decimation

Reply:
This is again for efficiency reasons.

Reply:
@teezeefpv here 4 seconds are used but the _crc_error_rate_factor is calculated per second. possible inconsistency

Reply:
Pretty sure this is just a pair of counters.